rust/ql/src/queries/unusedentities/UnreachableCode.ql

```diff
@@ -13,63 +13,60 @@ import codeql.rust.controlflow.ControlFlowGraph
 import codeql.rust.controlflow.internal.ControlFlowGraphImpl as ControlFlowGraphImpl
 
 /**
- * Holds if `n` is an AST node that's unreachable.
+ * Successor relation that includes unreachable AST nodes.
  */
-private predicate unreachable(AstNode n) {
-  not n = any(CfgNode cfn).getAstNode() and // reachable nodes
-  exists(ControlFlowGraphImpl::ControlFlowTree cft |
-    // nodes intended to be part of the CFG
-    cft.succ(n, _, _)
-    or
-    cft.succ(_, n, _)
-  )
+predicate succFull(AstNode a, AstNode b) {
+  exists(ControlFlowGraphImpl::ControlFlowTree cft | cft.succ(a, b, _))
 }
 
 /**
- * Holds if `n` is an AST node that's unreachable, and is not the successor
- * of an unreachable node (which would be a duplicate result).
+ * Gets a node we'd prefer not to report as unreachable.
  */
-private predicate firstUnreachable(AstNode n) {
-  unreachable(n) and
-  (
-    // no predecessor -> we are the first unreachable node.
-    not ControlFlowGraphImpl::succ(_, n, _)
-    or
-    // reachable predecessor -> we are the first unreachable node.
-    exists(AstNode pred |
-      ControlFlowGraphImpl::succ(pred, n, _) and
-      not unreachable(pred)
-    )
-  )
+predicate skipNode(AstNode n) {
+  // isolated node (not intended to be part of the CFG)
+  not succFull(n, _) and
+  not succFull(_, n)
```

Comment:
This bit confused me somewhat, because in most of the places where `skipNode` is used, this disjunct cannot possibly hold. The following rewrite should be equivalent:

```ql
private predicate succ(AstNode a, AstNode b) {
  exists(ControlFlowGraphImpl::ControlFlowTree cft | cft.succ(a, b, _))
}

private predicate succPostOrder(AstNode a, AstNode b) {
  succ(a, b) and
  a instanceof ControlFlowGraphImpl::PostOrderTree
}

/**
 * Successor relation that skips over `PostOrderTree`s.
 */
private predicate succSkipPostOrder(AstNode a, AstNode b) {
  exists(AstNode mid |
    not a instanceof ControlFlowGraphImpl::PostOrderTree and
    succ(a, mid) and
    succPostOrder*(mid, b) and
    not b instanceof ControlFlowGraphImpl::PostOrderTree
  )
}

/**
 * An AST node that is reachable.
 */
private predicate reachable(AstNode n) { n = any(CfgNode cfn).getAstNode() }

/**
 * Holds if `n` is an AST node that's unreachable, and any predecessors
 * of it are reachable (to avoid duplicate results).
 */
private predicate firstUnreachable(AstNode n) {
  // intended to be part of the CFG
  (
    succ(n, _) or
    succ(_, n)
  ) and
  not reachable(n) and
  not n instanceof ControlFlowGraphImpl::PostOrderTree and // location is counter-intuitive
  forall(AstNode pred | succSkipPostOrder(pred, n) | reachable(pred))
}
```

Reply:
The role of `skipNode` has grown and I failed to rename it.  Originally it was nodes where we'd rather report at a different location (for clarity), and there was a separate concept of isolated nodes we didn't want to report at all.  The logic was somewhat confusing and turned out to be faulty.

I've just done a few renamings including `skipNode` to `hiddenNode`.

In the new design, we're only doing one special case - producing a cut-down AST with all `hiddenNode`s removed.  They're linked over where appropriate, but for isolated nodes there are no links to make and they simply disappear.  That means we only need one mechanism to remove both `PostOrderTree` nodes and isolated nodes, and any other nodes we want to hide in future.